File: internal/disk/disk.go
//go:build windows

package disk

import (
	"encoding/json"
	"fmt"
	"os/exec"
	"strconv"
	"strings"
)

// Partition represents a disk partition
type Partition struct {
	Index       int    `json:"index"`
	DriveLetter string `json:"driveLetter"`
	Label       string `json:"label"`
	FileSystem  string `json:"fileSystem"`
	SizeBytes   uint64 `json:"sizeBytes"`
	SizeGB      string `json:"sizeGB"`
	IsBootable  bool   `json:"isBootable"`
	Type        string `json:"type"`
}

// DiskInfo represents a physical disk
type DiskInfo struct {
	Index      int         `json:"index"`
	Model      string      `json:"model"`
	SizeBytes  uint64      `json:"sizeBytes"`
	SizeGB     string      `json:"sizeGB"`
	Type       string      `json:"type"`
	Partitions []Partition `json:"partitions"`
}

// GetDisks retrieves disk and partition information using PowerShell
func GetDisks() ([]DiskInfo, error) {
	cmd := exec.Command("powershell", "-Command",
		"Get-Disk | Select-Object Number,Model,Size,PartitionStyle | ConvertTo-Json -Compress")
	out, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("failed to get disks: %w", err)
	}

	raw := strings.TrimSpace(string(out))
	if raw == "" {
		return []DiskInfo{}, nil
	}

	items, err := unmarshalPSJSON(raw)
	if err != nil {
		return nil, fmt.Errorf("failed to parse disk JSON: %w", err)
	}

	var disks []DiskInfo
	for i, item := range items {
		d := DiskInfo{Index: i}
		if v, ok := item["Number"].(json.Number); ok {
			if n, err := v.Int64(); err == nil {
				d.Index = int(n)
			}
		}
		if v, ok := item["Model"].(string); ok {
			d.Model = v
		}
		if v, ok := item["PartitionStyle"].(string); ok {
			d.Type = v
		}
		if v, ok := item["Size"].(json.Number); ok {
			if n, err := v.Int64(); err == nil {
				d.SizeBytes = uint64(n)
				d.SizeGB = FormatSize(d.SizeBytes)
			}
		}
		disks = append(disks, d)
	}

	return disks, nil
}

// GetPartitions retrieves all partitions with drive letters
func GetPartitions() ([]Partition, error) {
	cmd := exec.Command("powershell", "-Command",
		"Get-Volume | Where-Object {$_.DriveLetter} | Select-Object DriveLetter,FileSystemLabel,FileSystem,Size,SizeRemaining | ConvertTo-Json -Compress")
	out, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("failed to get partitions: %w", err)
	}

	raw := strings.TrimSpace(string(out))
	if raw == "" {
		return []Partition{}, nil
	}

	items, err := unmarshalPSJSON(raw)
	if err != nil {
		return nil, fmt.Errorf("failed to parse partition JSON: %w", err)
	}

	var partitions []Partition
	for i, item := range items {
		p := Partition{Index: i}
		if v, ok := item["DriveLetter"].(string); ok {
			p.DriveLetter = v
		}
		if v, ok := item["FileSystemLabel"].(string); ok {
			p.Label = v
		}
		if v, ok := item["FileSystem"].(string); ok {
			p.FileSystem = v
		}
		if v, ok := item["Size"].(json.Number); ok {
			if n, err := v.Int64(); err == nil {
				p.SizeBytes = uint64(n)
				p.SizeGB = FormatSize(p.SizeBytes)
			}
		}
		partitions = append(partitions, p)
	}

	return partitions, nil
}

// unmarshalPSJSON handles PowerShell's ConvertTo-Json output which may be a single object or an array
func unmarshalPSJSON(raw string) ([]map[string]interface{}, error) {
	decoder := json.NewDecoder(strings.NewReader(raw))
	decoder.UseNumber()

	// Try array first
	var arr []map[string]interface{}
	if err := decoder.Decode(&arr); err == nil {
		return arr, nil
	}

	// Try single object
	decoder = json.NewDecoder(strings.NewReader(raw))
	decoder.UseNumber()
	var obj map[string]interface{}
	if err := decoder.Decode(&obj); err != nil {
		return nil, err
	}
	return []map[string]interface{}{obj}, nil
}

// ValidateSpace checks if a partition has at least requiredBytes free
func ValidateSpace(driveLetter string, requiredBytes uint64) error {
	cmd := exec.Command("powershell", "-Command",
		fmt.Sprintf("(Get-Volume -DriveLetter %s).SizeRemaining", driveLetter))
	out, err := cmd.Output()
	if err != nil {
		return fmt.Errorf("failed to get free space: %w", err)
	}

	remainingStr := strings.TrimSpace(string(out))
	remaining, err := strconv.ParseUint(remainingStr, 10, 64)
	if err != nil {
		return fmt.Errorf("failed to parse free space: %w", err)
	}

	if remaining < requiredBytes {
		return fmt.Errorf("insufficient space: %d bytes required, %d available", requiredBytes, remaining)
	}

	return nil
}

// FormatSize formats bytes to human-readable GB
func FormatSize(bytes uint64) string {
	gb := float64(bytes) / (1024 * 1024 * 1024)
	return fmt.Sprintf("%.2f GB", gb)
}

File: src/disk.cpp
#include "disk.h"
#include <sstream>
#include <regex>

std::vector<Partition> GetPartitions(LogCallback log) {
    if (log) log(L"Enumerating partitions...");
    std::vector<Partition> result;

    std::wstring cmd = L"powershell.exe -NoProfile -Command \"Get-Volume | Where-Object {$_.DriveLetter} | Select-Object DriveLetter,FileSystemLabel,FileSystem,Size,SizeRemaining | ConvertTo-Json -Compress\"";

    std::string out;
    int exitCode = 0;
    if (!RunCommand(cmd, &out, &exitCode, log) || exitCode != 0) {
        if (log) log(L"  Failed to enumerate partitions.");
        return result;
    }

    // PowerShell JSON may be a single object or an array.
    // For simplicity, parse the key fields with regex since we control the output shape.
    std::regex objRegex(R"(\{([^}]*)\})");
    std::regex driveRegex(R"("DriveLetter"\s*:\s*"?([^",}]+)"?)");
    std::regex labelRegex(R"("FileSystemLabel"\s*:\s*"([^"]*)")");
    std::regex fsRegex(R"("FileSystem"\s*:\s*"([^"]*)")");
    std::regex sizeRegex(R"("Size"\s*:\s*(\d+))");
    std::regex freeRegex(R"("SizeRemaining"\s*:\s*(\d+))");

    std::string::const_iterator searchStart(out.cbegin());
    std::smatch m;
    while (std::regex_search(searchStart, out.cend(), m, objRegex)) {
        std::string block = m[1].str();
        Partition p;

        std::smatch dm;
        if (std::regex_search(block, dm, driveRegex)) {
            p.driveLetter = Utf8ToWide(Trim(dm[1].str()));
        }
        std::smatch lm;
        if (std::regex_search(block, lm, labelRegex)) {
            p.label = Utf8ToWide(lm[1].str());
        }
        std::smatch fm;
        if (std::regex_search(block, fm, fsRegex)) {
            p.fileSystem = Utf8ToWide(fm[1].str());
        }
        std::smatch sm;
        if (std::regex_search(block, sm, sizeRegex)) {
            p.sizeBytes = std::stoull(sm[1].str());
            p.sizeGB = FormatSize(p.sizeBytes);
        }
        std::smatch rm;
        if (std::regex_search(block, rm, freeRegex)) {
            p.freeBytes = std::stoull(rm[1].str());
        }

        if (!p.driveLetter.empty()) {
            result.push_back(p);
        }
        searchStart = m.suffix().first;
    }

    if (log) log(L"  Found " + std::to_wstring(result.size()) + L" partition(s).");
    return result;
}

bool ValidateSpace(const std::wstring& driveLetter, uint64_t requiredBytes, LogCallback log) {
    if (log) log(L"Checking free space on " + driveLetter + L":...");
    std::wostringstream cmd;
    cmd << L"powershell.exe -NoProfile -Command \"(Get-Volume -DriveLetter " << driveLetter << L").SizeRemaining\"";

    std::string out;
    int exitCode = 0;
    if (!RunCommand(cmd.str(), &out, &exitCode, log) || exitCode != 0) {
        if (log) log(L"  Failed to query free space.");
        return false;
    }

    std::string num = Trim(out);
    if (num.empty()) {
        if (log) log(L"  Error: could not parse free space.");
        return false;
    }
    try {
        uint64_t remaining = std::stoull(num);
        if (remaining < requiredBytes) {
            if (log) {
                log(L"  Insufficient space: required " + Utf8ToWide(FormatSize(requiredBytes)) +
                    L", available " + Utf8ToWide(FormatSize(remaining)));
            }
            return false;
        }
        if (log) log(L"  Space OK.");
        return true;
    } catch (...) {
        if (log) log(L"  Error: failed to parse free space value.");
        return false;
    }
}
